src/main/java/org/broadinstitute/hellbender/engine/dataflow/transforms/PairReadWithRefBases.java

```diff
@@ -0,0 +1,51 @@
+package org.broadinstitute.hellbender.engine.dataflow.transforms;
+
+import com.google.cloud.dataflow.sdk.transforms.DoFn;
+import com.google.cloud.dataflow.sdk.transforms.PTransform;
+import com.google.cloud.dataflow.sdk.transforms.ParDo;
+import com.google.cloud.dataflow.sdk.values.KV;
+import com.google.cloud.dataflow.sdk.values.PCollection;
+import org.broadinstitute.hellbender.utils.SimpleInterval;
+import org.broadinstitute.hellbender.utils.read.GATKRead;
+import org.broadinstitute.hellbender.utils.reference.ReferenceBases;
+
+/**
+ * PairReadWithRefBases pairs the ReferenceBases overlapping the reads with each read and produces KV with read as
+ * the key.
+ *
+ * |---------- ref bases 1 ---------|      |-------------------- ref bases 2 --------------------|
+ * |-- read a --|   |---- read b ---|         |--------------- read c --------------|
+ * results in,
+ *
+ * |-- read a --|   |---- read b ---|         |--------------- read c --------------|
+ * |ref bases 1a|   |-ref bases 1b -|         |----------- ref bases 2c ------------|
+ *
+ * KV<read a, ref bases 1a>
+ * KV<read b, ref bases 1b>
+ * KV<read c, ref bases 2c>
+ * Precondition: each refbases must contain ALL of the bases for all of the GATKReads it's paired with.
+ */
+
+public class PairReadWithRefBases extends PTransform<PCollection<KV<ReferenceBases, Iterable<GATKRead>>>, PCollection<KV<GATKRead, ReferenceBases>>> {
+    private static final long serialVersionUID = 1L;
+
+    @Override
+    public PCollection<KV<GATKRead, ReferenceBases>> apply(PCollection<KV<ReferenceBases, Iterable<GATKRead>>> input) {
+        return input.apply(ParDo.of(new DoFn<KV<ReferenceBases, Iterable<GATKRead>>, KV<GATKRead, ReferenceBases>>() {
+            private static final long serialVersionUID = 1L;
+            @Override
+            public void processElement(ProcessContext c) throws Exception {
```

Comment:
Serialization/Deserialization costs can easily dominate, slowing down a dataflow program. If we're amplifying a 3GB input to 150GB of data, we may find that our program spends a lot of time on I/O. This is especially problematic if the amplified data ever has to travel from one machine to another.

One way of being "careful" would be for example to make sure that this amplified data never needs to be serialized (it stays within the same machine and we don't hit any perf bug in the dataflow runtime). Another way to be careful is to make sure we don't generate more data than fits in the machine's memory - for example by having multiple objects share some of their (immutable) data.